src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs

```diff
@@ -23,6 +24,30 @@
         => new(value);
 }
 
+internal sealed class AsyncLazy<T> : AsyncLazy<
+    T,
+    (Func<CancellationToken, Task<T>> asynchronousComputeFunction, Func<CancellationToken, T>? synchronousComputeFunction)>
+{
+    public AsyncLazy(T value) : base(value)
+    {
+    }
+
+    public AsyncLazy(Func<CancellationToken, Task<T>> asynchronousComputeFunction)
```

Comment:
Something feels a bit off, for example I added the following method to AsyncLazy:

    public static AsyncLazy<T, TData> Create<T, TData>(Func<TData, CancellationToken, Task<T>> asynchronousComputeFunction, TData data)
        => new(asynchronousComputeFunction, data);

and then I switched SolutionCompilationState._lazyChecksums over by making the following changes. It felt wrong to include the data type in the declaration

``` 
// original code
private AsyncLazy<SolutionCompilationStateChecksums> _lazyChecksums;

private SolutionCompilationState(...)
{
    _lazyChecksums = AsyncLazy.Create(c => ComputeChecksumsAsync(projectId: null, c));
}
```

and I needed to specify the data type in the member declaration and wasn't able to use the Create

``` 
// new code
private readonly AsyncLazy<SolutionCompilationStateChecksums, SolutionCompilationState> _lazyChecksums;

private SolutionCompilationState(...)
{
    _lazyChecksums = AsyncLazy.Create(static (self, c) => self.ComputeChecksumsAsync(projectId: null, c), data: this);
}
```

Reply:
is it any better if you reverse the order of the arguments?